### ISSUE TITLE
Firmware bug causing spontaneous shutdown on SP7 (and probably other Surface devices)

### ISSUE THREAD

**Issue by quo** [linux-surface/linux-surface#999]
The Surface Pro 7 has a firmware bug that causes the SAM EC to reset every 49.7 days (1<<32 milliseconds). When the SAM resets, the Surface Pro powers off.

The timer is not reset by normal shutdowns, as the SAM will keep running as long as the battery is not empty. So the bug can trigger even if you've just turned on the tablet, if the SAM has already been running for 49 days since its last reset. The only way to reset the timer is to reset the SAM, by holding the power and volume up buttons for at least 15 seconds, or draining the battery.

The bug doesn't always trigger, but the odds of it triggering are very high. I've only seen one case where a SP7 survived past 49.7 days, and it then shut down after the next 49.7 day interval. 

The bug is still present in SP7 firmware version 14.502.139. The same bug is also present in the firmware of the SP7+ (version 14.102.139), SP8 (version 10.8.139), and SP9 (version 6.89.139), and probably other Surface devices, although I have not verified that it causes the same symptoms on these devices.

We also use a number of SP4s, SP5s and SP6s, and we have not seen this issue on those devices. I have not yet looked at the firmware for them to see if the bug is completely absent, or if there is some other reason it doesn't trigger.

I have created a firmware patch to fix the problem on the SP7.

# Details

The SAM firmware internally uses a 64-bit microsecond timer. It divides this by 1000 to create 64-bit millisecond timer which is used for timeouts, etc.

The bug is in a function that schedules sleeping tasks in the SAM RTOS ("Kaos"). The function is found at 0x83de in the latest 14.502.139 firmware. This function takes a single parameter, a task handle. It loops through the linked list of sleeping tasks, and inserts the task in the right place, so that the first task to wake up will be at the front of the list. The problem is that this function calculates the wake time for each task as a 32-bit value. It effectively does:

```c
uint32_t wake_time = UINT32_MAX;
if (task[i].timeout != UINT32_MAX)
  wake_time = (uint32_t)task[i].timestamp + task[i].timeout;
```

There is also a watchdog task which looks like this:

```c
watchdog_init_ms(20000);
while(1) {
	sleep_ms(1000);
	watchdog_feed();
}
```

So what happens is that when the calculated wake time overflows 32 bits, a task is incorrectly placed at the front of the list, and blocks other tasks from waking at the right time. This causes the watchdog thread to hang in its sleep call, and after several seconds the watchdog triggers and resets the SAM, turning off the Surface Pro.

The problem is of course easily fixed by using 64-bit arithmetic, which is what my patch does:

```c
uint64_t wake_time = UINT64_MAX;
if (task[i].timeout != UINT32_MAX)
  wake_time = task[i].timestamp + task[i].timeout;
```

The same bug is also present in the function that schedules timers (at 0x8f90). It also uses 32-bit instead of 64-bit arithmetic.

# Fix

I have tried to contact someone at MS who has done Surface firmware work, but I haven't received a response yet. If they stay silent, I suppose I'll try the official support channels (and if anyone else has any contacts at MS, feel free to forward this to them). Once MS confirm the bug I imagine they'll publish firmware updates for all affected devices, but it will probably take a while.

In the mean time, I have developed my own patch for the SP7 firmware:
https://github.com/quo/sam-fw-tools/tree/master/bugfix

While I have successfully used this patch, please note that:
- If you install this, it is entirely AT YOUR OWN RISK. I am NOT responsible if it breaks your device or causes any other kind of damage.
- I CANNOT help you if something goes wrong. If your SP7 breaks, your only option is probably to send it back to MS for repair.

That said, the steps to apply the fix are as follows:
- Download the offical firmware pack from Microsoft.
- Extract the files using `msiextract`.
- Go to the directory containing the extracted SAM firmware: `cd SurfaceUpdate/surfacesam`
- Get the SAM FW scripts: `git clone https://github.com/quo/sam-fw-tools`
- Unpack the firmware: `./sam-fw-tools/sam-fw-unpack.py SurfaceSAM_14.502.139.bin`
- Patch the two firmware images: `for x in SurfaceSAM_14.502.139.bin.?.img; do ./sam-fw-tools/sam-fw-patch.py $x sam-fw-tools/bugfix/SP7bugfix.14.502.139.patch; done`
- Make sure the [SAM kernel modules](https://github.com/linux-surface/surface-aggregator-module) are installed and load the cdev client: `sudo modprobe surface_aggregator_cdev`
- Upload the patched firmware: `sudo ./sam-fw-tools/sam-fw-upload.py SurfaceSAM_14.502.139.bin.0.img.patched`
- If the previous command gives the error "cannot flash active firmware location", you'll need to upload the other patched image: `sudo ./sam-fw-tools/sam-fw-upload.py SurfaceSAM_14.502.139.bin.1.img.patched`
- Once the firmware has been uploaded, reset the SAM by holding the power and volume up buttons for at least 15 seconds (make sure to hold them even if the screen goes black). This will turn off the tablet, so to avoid data loss/corruption in Linux it's probably best to do it from the UEFI menu.
- The patch changes the firmware version from 14.502.139 to 14.502.140, so you can verify that it has been installed from the UEFI menu.

**Comment by quo** [comment]
I've now had a (brief) look at the SP6 firmware. It looks like the bug is not present there, because it does not use a separate ordered list of waiting tasks. Instead it just scans through all the tasks when switching tasks. And that code correctly uses 64 bit comparisons.

So the list of waiting tasks was added in the SP7 firmware as an optimization, but unfortunately with a serious bug...

It looks like the buggy code is present in the SAM firmware of the following devices (so **these are probably all affected**):
- Surface Pro 7/7+/8/9
- Surface Laptop 3/4/5
- Surface Laptop Studio
- Surface Book 3
- Surface Go / Go 2
- Surface Hub 2S
- Surface Studio 2+

Whereas the original non-optimized code is present in the firmware of the following devices (so **these are not affected**):
- Surface Pro 4/5/6
- Surface Laptop 1/2
- Surface Book 1/2
- Surface Studio 2

(For remaining devices, I was not able to obtain the firmware, or they use a different encoding.)

**Comment by quo** [comment]
Now reported to MS at: https://answers.microsoft.com/en-us/surface/forum/all/sp7-spontaneously-powers-off-due-to-firmware-bug/f75b5d74-42b1-4278-babf-0374e549605d

**Comment by quo** [comment]
Looks like this bug was fixed in SP7 SAM firmware version 14.701.139.0, released September 2023.

Coincidentally my firmware upload script also doesn't work anymore with the current firmware... ;)
